rclcpp_components/src/component_manager.cpp

```diff
@@ -54,9 +73,19 @@ ComponentManager::~ComponentManager()
     RCLCPP_DEBUG(get_logger(), "Removing components from executor");
     if (auto exec = executor_.lock()) {
       for (auto & wrapper : node_wrappers_) {
+        // if node is in dedicated executors, ignore it.
+        if (dedicated_executors_.find(wrapper.first) != dedicated_executors_.end()) {
+          continue;
+        }
         exec->remove_node(wrapper.second.get_node_base_interface());
       }
     }
+    for (auto & exec : dedicated_executors_) {
+      exec.second->cancel();
```

Comment:
https://github.com/ros2/rclcpp/blob/5fcb7892f70bd741cf86b019fc489f3a81a60510/rclcpp_components/src/component_manager.cpp#L257-L258

It is added to `dedicated_executors_` right before being spun, so it should always have been called `spin` to be on that map. I suppose the only way to make it more bulletproof is if those 2 lines are swapped so that its only added to `dedicated_executors_` _after_ its been spinning.

Reply:
Swapping the lines doesn't help.

The issue is that though the thread is already spawned, you don't actually know when there will be a context switch and `spin()` will be actually called.
In an extremely unlucky scenario, the thread is spawn and an unload request is received before `spin()` is called, resulting in the cancellation request being lost.

the "right way" to do it is:

```cpp
std::promise<void> prom;
dedicated_threads_[node_id] = std::thread([exec, cancel_token=prom->get_future()]() {exec->spin_until_future_complete(cancel_token);});
dedicated_promises_[node_id] = std::move(promise);
```

```
-      dedicated_executor->second->cancel();
+      dedicated_promise->second->set_value();
```

(the code will need some more changes than that, that's just the idea)

there's no race possible in that case, as a future/promise channel is stateful.

Reply:
Thanks for the specific suggestion!

Reply:
thanks for more details about how to implement, it's helpful to me.